S3 error: You must provide the Content-Length HTTP header

When pushing Docker image, I got 500 error from stacktrace from S3 client `You must provide the Content-Length HTTP header`. It seems the request contains `Content-Length` header, but it was not provided to S3 storage:
```
artipie    | 
artipie    | [INFO] 2020-06-05 14:00:58,800 vert.x-eventloop-thread-1 com.artipie.http.async.AsyncSlice - >> PATCH /public/docker/v2/alpine/blobs/uploads/4741e16d-3190-4391-a925-4b40cb3dd957 HTTP_1_0
artipie    | 
artipie    | Host: central.artipie.com
artipie    | X-Real-IP: x.x.x.x
artipie    | X-Forwarded-For: x.x.x.x
artipie    | Connection: close
artipie    | Content-Length: 2757034
artipie    | User-Agent: docker/19.03.8...
artipie    | Accept-Encoding: gzip
artipie    | [INFO] 2020-06-05 14:00:59,078 sdk-async-response-11-0 com.artipie.asto.s3.S3Storage - List 'docker/docker/registry/v2/repositories/alpine/_uploads/4741e16d-3190-4391-a925-4b40cb3dd957': 1
artipie    | [INFO] 2020-06-05 14:00:59,140 sdk-async-response-11-1 com.artipie.asto.s3.S3Storage - Size 'docker/docker/registry/v2/repositories/alpine/_uploads/4741e16d-3190-4391-a925-4b40cb3dd957/data': 0
artipie    | [INFO] 2020-06-05 14:00:59,219 sdk-async-response-11-2 com.artipie.http.async.AsyncSlice - Failure: java.util.concurrent.CompletionException: software.amazon.awssdk.services.s3.model.S3Exception: You must provide the Content-Length HTTP header. (Service: S3, Status Code: 411, Request ID: 0000017284C8EA1D8019A0141AC32C96)
artipie    | 	at software.amazon.awssdk.utils.CompletableFutureUtils.errorAsCompletionException(CompletableFutureUtils.java:61)
artipie    | 	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncExecutionFailureExceptionReportingStage.lambda$execute$0(AsyncExecutionFailureExceptionReportingStage.java:51)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.uniHandle(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
artipie    | 	at software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0(CompletableFutureUtils.java:75)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
artipie    | 	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.retryResponseIfNeeded(AsyncRetryableStage.java:156)
artipie    | 	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.retryIfNeeded(AsyncRetryableStage.java:120)
artipie    | 	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.lambda$execute$0(AsyncRetryableStage.java:107)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)
artipie    | 	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$executeHttpRequest$1(MakeAsyncHttpRequestStage.java:138)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
artipie    | 	at java.base/java.lang.Thread.run(Unknown Source)
artipie    | Caused by: software.amazon.awssdk.services.s3.model.S3Exception: You must provide the Content-Length HTTP header. (Service: S3, Status Code: 411, Request ID: 0000017284C8EA1D8019A0141AC32C96)
artipie    | 	at software.amazon.awssdk.services.s3.model.S3Exception$BuilderImpl.build(S3Exception.java:95)
artipie    | 	at software.amazon.awssdk.services.s3.model.S3Exception$BuilderImpl.build(S3Exception.java:55)
artipie    | 	at software.amazon.awssdk.protocols.query.unmarshall.AwsXmlErrorProtocolUnmarshaller.handle(AwsXmlErrorProtocolUnmarshaller.java:127)
artipie    | 	at software.amazon.awssdk.protocols.query.unmarshall.AwsXmlErrorProtocolUnmarshaller.handle(AwsXmlErrorProtocolUnmarshaller.java:86)
artipie    | 	at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.lambda$prepare$0(AsyncResponseHandler.java:88)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
artipie    | 	at java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)
artipie    | 	at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler$BaosSubscriber.onComplete(AsyncResponseHandler.java:129)
artipie    | 	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.runAndLogError(ResponseHandler.java:180)
artipie    | 	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.access$500(ResponseHandler.java:67)
artipie    | 	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$PublisherAdapter$1.onComplete(ResponseHandler.java:296)
artipie    | 	at com.typesafe.netty.HandlerPublisher.publishMessage(HandlerPublisher.java:362)
artipie    | 	at com.typesafe.netty.HandlerPublisher.flushBuffer(HandlerPublisher.java:304)
artipie    | 	at com.typesafe.netty.HandlerPublisher.receivedDemand(HandlerPublisher.java:258)
artipie    | 	at com.typesafe.netty.HandlerPublisher.access$200(HandlerPublisher.java:41)
artipie    | 	at com.typesafe.netty.HandlerPublisher$ChannelSubscription$1.run(HandlerPublisher.java:452)
artipie    | 	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
artipie    | 	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
artipie    | 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
artipie    | 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
artipie    | 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
artipie    | 	... 1 more
artipie    | 
```

--- a/src/main/java/com/artipie/asto/s3/S3Storage.java
+++ b/src/main/java/com/artipie/asto/s3/S3Storage.java
@@ -161,9 +161,22 @@ public final class S3Storage implements Storage {
     public CompletableFuture<Void> save(final Key key, final Content content) {
         final CompletableFuture<Void> result;
         if (this.multipart) {
-            result = this.saveWithMultipartAllowed(key, content);
+            result = complementWithSize(content, S3Storage.MIN_MULTIPART).thenCompose(
+                updated -> {
+                    final CompletableFuture<Void> future;
+                    final Optional<Long> size = updated.size();
+                    if (size.isPresent() && size.get() < S3Storage.MIN_MULTIPART) {
+                        future = this.put(key, updated);
+                    } else {
+                        future = this.uploadMultipart(key, updated);
+                    }
+                    return future;
+                }
+            );
         } else {
-            result = this.put(key, content);
+            result = complementWithSize(content).thenCompose(
+                updated -> this.put(key, updated)
+            );
         }
         return result;
     }
@@ -242,50 +255,61 @@ public final class S3Storage implements Storage {
     }
 
     /**
-     * Saves the bytes to the specified key.
-     * Uses multipart upload feature if size is greater or equal MIN_MULTIPART.
+     * Complements {@link Content} with size if size is unknown.
+     * Size calculated by reading all content bytes.
      *
-     * @param key The key
-     * @param content Bytes to save
-     * @return Completion or error signal.
+     * @param content Original content.
+     * @return Updated content with size.
      */
-    private CompletableFuture<Void> saveWithMultipartAllowed(final Key key, final Content content) {
-        final Flowable<ByteBuffer> cache = Flowable.fromPublisher(content).cache();
+    private static CompletableFuture<Content> complementWithSize(final Content content) {
+        return complementWithSize(content, Long.MAX_VALUE);
+    }
+
+    /**
+     * Complements {@link Content} with size if size is unknown.
+     * Size calculated by reading up to `limit` content bytes.
+     * If end of content has not been reached by reading `limit` of bytes
+     * then original content is returned.
+     *
+     * @param content Original content.
+     * @param limit Content reading limit.
+     * @return Updated content with size.
+     * @todo #183:60min Read no more then limit from Content.
+     *  Now method does not fulfil it's contract and caches all content bytes to `Flowable`,
+     *  then reads from them up to the limit.
+     *  This should be fixed, so no more then `limit` bytes is cached.
+     */
+    private static CompletableFuture<Content> complementWithSize(
+        final Content content,
+        final long limit
+    ) {
         return content.size()
-            .map(size -> CompletableFuture.completedFuture(Optional.of(size)))
+            .map(ignored -> CompletableFuture.completedFuture(content))
             .orElseGet(
-                () -> cache
-                    .map(Buffer::remaining)
-                    .scanWith(() -> 0L, (sum, item) -> sum + item)
-                    .takeUntil(total -> total >= S3Storage.MIN_MULTIPART)
-                    .lastOrError()
-                    .to(SingleInterop.get())
-                    .toCompletableFuture()
-                    .<Optional<Long>>thenApply(
-                        last -> {
-                            final Optional<Long> size;
-                            if (last >= S3Storage.MIN_MULTIPART) {
-                                size = Optional.empty();
-                            } else {
-                                size = Optional.of(last);
+                () -> {
+                    final Flowable<ByteBuffer> cache = Flowable.fromPublisher(content).cache();
+                    return cache
+                        .map(Buffer::remaining)
+                        .scanWith(() -> 0L, (sum, item) -> sum + item)
+                        .takeUntil(total -> total >= limit)
+                        .lastOrError()
+                        .to(SingleInterop.get())
+                        .toCompletableFuture()
+                        .<Optional<Long>>thenApply(
+                            last -> {
+                                final Optional<Long> size;
+                                if (last >= limit) {
+                                    size = Optional.empty();
+                                } else {
+                                    size = Optional.of(last);
+                                }
+                                return size;
                             }
-                            return size;
-                        }
-                    )
-            ).thenApply(
-                sizeOpt -> sizeOpt
-                    .<Content>map(size -> new Content.From(size, cache))
-                    .orElse(new Content.From(cache))
-            ).thenCompose(
-                updated -> {
-                    final CompletableFuture<Void> future;
-                    final Optional<Long> size = updated.size();
-                    if (size.isPresent() && size.get() < S3Storage.MIN_MULTIPART) {
-                        future = this.put(key, updated);
-                    } else {
-                        future = this.uploadMultipart(key, updated);
-                    }
-                    return future;
+                        ).thenApply(
+                            sizeOpt -> sizeOpt
+                                .<Content>map(size -> new Content.From(size, cache))
+                                .orElse(new Content.From(cache))
+                        );
                 }
             );
     }

